dagrunner/plugin_framework.py

```diff
@@ -59,6 +64,105 @@ def __call__(self, *args, **kwargs):
         return subprocess.run(*args, **kwargs, shell=True, check=True)
 
 
+def _stage_to_dir(*args, staging_dir, verbose=False):
+    """
+    Copy input filepaths to a staging area and update paths.
+
+    Hard link copies are preferred (same host) and physical copies are made otherwise.
+    File name, size and modification time are used to evaluate if the destination file
+    exists already (matching criteria of rsync).  If exists already, skip the copy.
```

Comment:
If both the file size and modification time are the same, `rsync` assumes the files are identical and does not transfer them. This default behaviour is usually sufficient for many use cases and is faster because it avoids the overhead of calculating checksums.